test/YesSql.Tests/CoreTests.cs

```diff
@@ -4155,6 +4155,8 @@ public async Task ShouldQueryInnerSelectWithCollection()
                 Assert.Equal(0, await session.Query<Person, PersonByNameCol>(collection: "Col1").Where(x => x.Name.IsNotInAny<PersonByBothNamesCol>(y => y.Firstname)).CountAsync());
 
                 Assert.Equal(2, await session.Query("Col1").For<Person>().With<PersonByNameCol>().Where(x => x.Name.IsInAny<PersonByBothNamesCol>(y => y.Firstname)).CountAsync());
+                Assert.Equal(2, await session.Query().For<Person>(collection:"Col1").With<PersonByNameCol>().Where(x => x.Name.IsInAny<PersonByBothNamesCol>(y => y.Firstname)).CountAsync());
+                Assert.Equal(2, await session.Query("Col2").For<Person>(collection: "Col1").With<PersonByNameCol>().Where(x => x.Name.IsInAny<PersonByBothNamesCol>(y => y.Firstname)).CountAsync());
```

Comment:
Actually I am confused now. `Query` already has a constructor with a collection, so the `For` call should use this one. So I read the issue again and I see that you are trying to hack the state of the query. I'd prefer a solution where you can plug a custom "CollectionResolver" that could be configured instead. So the `Query()` constructor would either take a collection name, or a `CollectionResolver`, or nothing to use default Collectionresolver from the configuration. That might be even better for your design.